### PR TITLE
fix Issue 24084 - Add -nothrow Switch to Compiler

### PR DESCRIPTION
**Pull request by WalterBright**
@ibuclaw suggested this, gdc does it.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -21,6 +21,7 @@ dub.selections.json
 /compiler/src/vcbuild/Win32
 /compiler/src/vcbuild/x64
 /compiler/src/vcbuild/.vs
+/compiler/src/bug2
```
Comment (dkorpel) [inline .gitignore:24]: What's this for?

Comment (WalterBright) [reply .gitignore:24]: It's a junk directory on my system that git suddenly started complaining about.

Comment (maxhaton) [reply .gitignore:24]: `git add -p` is your friend

**Comment by maxhaton** [comment]
Isn't this just more more flag-based balkanization?

**Comment by maxhaton** [comment]
Unless you complete change the ABI I don't see how it's going to much more efficient other than emitting less tables into the binary anyway.

**Comment by WalterBright** [comment]
> I don't see how it's going to much more efficient

Optimizations that cross handler lines don't get done.

File: .gitignore (hunk 1 above)
Comment (etienne02) [reply .gitignore:24]: Otherwise, you can also add it to your `.git/info/exclude` this is a gitignore for you only (not shared)

**Comment by ibuclaw** [comment]
> Unless you complete change the ABI I don't see how it's going to much more efficient other than emitting less tables into the binary anyway.

And emitting less compiler-generated field dtors - as well as the EH regions emitted around constructors with auto-dtor fields.

File: .gitignore (hunk 1 above)
Comment (RazvanN7) [reply .gitignore:24]: @WalterBright please take care of this so we can merge this.

**Review by RazvanN7 — DISMISSED** [review]
@WalterBright Please delete de unused directory and add a changelog entry.

**Comment by maxhaton** [comment]
> > Unless you complete change the ABI I don't see how it's going to much more efficient other than emitting less tables into the binary anyway.
> 
> And emitting less compiler-generated field dtors - as well as the EH regions emitted around constructors with auto-dtor fields.

Example?

**Comment by WalterBright** [comment]
> Example?

Variables that cross exception lines cannot be put in registers, as there's no way to restore them when transferring control to the handler.

**Comment by maxhaton** [comment]
> > Example?
> 
> Variables that cross exception lines cannot be put in registers, as there's no way to restore them when transferring control to the handler.

If there is no exceptions in the first place how does this flag help with that? LLVM and GCC can both do data flow analysis across exceptional control flow for example.

**Comment by WalterBright** [comment]
> If there is no exceptions in the first place how does this flag help with that?

Register allocation is done at compile time, not runtime. This switch tells the compiler that no exceptions will be thrown, so it can generate code as if no exceptions were thrown.

**Comment by maxhaton** [comment]
> > If there is no exceptions in the first place how does this flag help with that?
> 
> Register allocation is done at compile time, not runtime. This switch tells the compiler that no exceptions will be thrown, so it can generate code as if no exceptions were thrown.

But for the code to compile with the flag in the first place where would the exceptional control flow effect register allocation? There couldn't be any throw, try, or catch statements in the first place. The whole point of (say) DWARF exceptions is to shove all the (time) cost of using them into to the support routines. The only other places I can think of it really having an effect around regardless of the compiler's ability to work around exceptions or not in the presence of an expensive function call that will put a bunch of constraints on the register allocator anyway.

**Comment by maxhaton** [comment]
Also what happens if I call `new` and the allocation fails?

Based on an experiment (turning useExceptions to false then recompiling): 
```
src/rt/dwarfeh.d:340: uncaught exception reached top of stack
This might happen if you're missing a top level catch in your fiber or signal handler
```
 i.e. not great. The compiler can and does allocate enough memory for this to happen

**Comment by maxhaton** [comment]
Takeaway of the above being that the flag itself has uses i.e. I think betterC should go but keeping its constituent options exposed is reasonable, but the compiler probably shouldn't use it.

**Comment by WalterBright** [comment]
> for the code to compile with the flag in the first place where would the exceptional control flow effect register allocation?

When you have a variable declared outside the try-catch block but used within the try-catch block. It cannot be placed in a register.

> There couldn't be any throw, try, or catch statements in the first place.

RAII objects are re-written into try-finally constructs. So are scope statements.

> what happens if I call new and the allocation fails?

fatal error

Exceptions have always been sold as zero-cost if no exceptions are thrown. But it's not true. gcc and clang have a `-fno-exceptions` flag, in VC `/EH` turns exceptions on (it defaults to off) and in DMC++ `/Ae` turns them on. @ibuclaw says gdc is compiled with exceptions off.

**Comment by ibuclaw** [comment]
> Exceptions have always been sold as zero-cost if no exceptions are thrown. But it's not true. gcc and clang have a `-fno-exceptions` flag, in VC `/EH` turns exceptions on (it defaults to off) and in DMC++ `/Ae` turns them on. @ibuclaw says gdc is compiled with exceptions off.

The rationale for that is because the `main` function for gdc is defined in C++ and controlled by gcc, not us. The whole of the C++ codebase in gcc is compiled with `-fno-exceptions`. By also compiling the D code with exceptions off helps us - for want of a better word - be a "better C++" compiler in this project.

**Comment by ibuclaw** [comment]
> > > Unless you complete change the ABI I don't see how it's going to much more efficient other than emitting less tables into the binary anyway.
> > 
> > And emitting less compiler-generated field dtors - as well as the EH regions emitted around constructors with auto-dtor fields.
> 
> Example?

Compiling gdc code with D exceptions on
```
77M     gcc/d21 
```
And same but with exceptions off 
```
51M     gcc/d21
```

**Review by PetarKirov — APPROVED** [review]
Very nice step in the right direction. I'm happy to see `-betterC` being broken off into smaller more modular parts.

**Pull request merged**